python/ray/train/base_trainer.py

```diff
@@ -164,8 +175,68 @@ def __init__(
         self.preprocessor = preprocessor
         self.resume_from_checkpoint = resume_from_checkpoint
 
+        # This path is only set through the `restore` path of constructing the trainer
+        self._restore_path = None
+
         self._validate_attributes()
 
+    @classmethod
+    def restore(
+        cls: Type["BaseTrainer"],
+        path: str,
+        datasets: Optional[Dict[str, GenDataset]] = None,
+        preprocessor: Optional["Preprocessor"] = None,
+        scaling_config: Optional[ScalingConfig] = None,
+        **kwargs,
+    ) -> "BaseTrainer":
+        trainer_state_path = cls._maybe_sync_down_trainer_state(path)
+        assert trainer_state_path.exists()
+
+        with open(trainer_state_path, "rb") as fp:
+            trainer = pickle.load(fp)
+        assert type(trainer) == cls
+
+        invalid_fields = [
+            k for k in kwargs if k not in cls._optional_fields_for_restore
+        ]
+        valid_fields = [
+            "datasets",
+            "preprocessor",
+            "scaling_config",
+        ] + cls._optional_fields_for_restore
+        assert not invalid_fields, (
+            f"`{cls.__name__}.restore(path, ...)` only takes in the following "
+            "set of fields as additional arguments:\n"
+            f"\t{valid_fields}\n"
+            f"You passed in the following invalid arguments: {invalid_fields}"
+        )
+
+        trainer._restore_path = path
+
+        if trainer.datasets and not datasets:
+            raise ValueError()
+        original_datasets = trainer.datasets or {}
+        datasets = datasets or {}
+        assert set(original_datasets.keys()) == set(datasets.keys())
+        trainer.datasets = datasets
+
+        # If no preprocessor is re-specified, then it will be set to None
+        # here and loaded from the latest checkpoint
+        trainer.preprocessor = preprocessor
+
+        if scaling_config:
+            trainer.scaling_config = scaling_config
+
+        for field_name in cls._optional_fields_for_restore:
+            original_attr = getattr(trainer, field_name, None)
+            respecified = field_name in kwargs
+            if contains_object_refs(original_attr) and not respecified:
+                raise ValueError()
```

Comment:
We probably want to fill this out

Reply:
Yep, going to do a pass to fill in all the errors!